xmake/rules/platform/windows/idl/xmake.lua

```diff
@@ -42,26 +70,130 @@ rule("platform.windows.idl")
             "/proxy",  name .. "_p.c",
             "/tlb",    name .. ".tlb",
+            "/cstub",  name .. "_c.c",
+            "/sstub",  name .. "_s.c",
+            "/server", (enable_server and "stub" or "none"),
+            "/client", (enable_client and "stub" or "none"),
             path(sourcefile)
         })
 
-        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.idl %s", sourcefile)
-        batchcmds:vrunv(midl.program, flags, {envs = msvc:runenvs()})
+        -- use this way to set depend to avoid generating files multiple times
+        depend.on_changed(function() 
+            progress.show(opt.progress, "${color.build.object}generating.idl %s", sourcefile)
+            os.vrunv(midl.program, flags, { envs = msvc:runenvs() })
+        end, { files = sourcefile, dependfile = path.join(autogendir, path.basename(sourcefile) .. ".idl.d") })
+    end)
+    
+    --[[
+        we don't have a way to detect which files midl.exe has generated and os.exists
+        does not work in before_buildcmd_file because in the invokation of xmake
+        the files might not have been generated yet from batchcmds, therefore
+        the files are compiled and checked during the buildcmd as _i, _p, _c, _s
+        might not exists depending on the idl file
+    ]]
+    on_buildcmd_file(function (target, batchcmds, sourcefile, opt)
+        import("core.project.depend")
+
+        local name = path.basename(sourcefile)
+        local autogendir = path.join(target:autogendir(), "platform/windows/idl")
+
+        -- we don't have a way to detect which midl files are generated
+
+        local icfile = path.join(autogendir, name .. "_i.c")
+        local icobj = target:objectfile(icfile)
 
-        local iid_file = path.join(autogendir, name .. "_i.c")
-        local objectfile = target:objectfile(iid_file)
-        table.insert(target:objectfiles(), objectfile)
+        local scfile = path.join(autogendir, name .. "_s.c")
+        local scobj = target:objectfile(scfile)
+
+        local ccfile = path.join(autogendir, name .. "_c.c")
+        local ccobj = target:objectfile(ccfile)
+
+        local pcfile = path.join(autogendir, name .. "_p.c")
+        local pcobj = target:objectfile(pcfile)
+
+        local fileconfig = target:fileconfig(sourcefile)
+        local enable_proxy = true
+        if fileconfig then
+            if fileconfig.proxy ~= nil then
+                enable_proxy = fileconfig.proxy
+            end
+        end
 
-        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.$(mode) %s", iid_file)
-        batchcmds:compile(iid_file, objectfile)
+        -- compile c files
+        local configs = {includedirs = autogendir, languages = "c89"}
 
-        batchcmds:add_depfiles(sourcefile, iid_file)
-        batchcmds:set_depmtime(os.mtime(objectfile))
-        batchcmds:set_depcache(target:dependfile(objectfile))
+        if os.exists(icfile) then
```

Comment:
we should not use it, please remove os.exitss

Reply:
we don't have a way to detect which files midl.exe has generated and os.exists
        does not work in before_buildcmd_file because in the invokation of xmake
        the files might not have been generated yet from batchcmds, therefore
        the files are compiled and checked during the buildcmd as _i, _p, _c, _s
        might not exists depending on the idl file, so I had to use os.exists, any idea to solve this if I can't use that?

Reply:
Using batchcmds has some limitations and can usually only be used for simple build rules and project generators.

If we want to handle them dynamically better, we should use jobgraph + on_build_file + depend.on_changed

see

https://github.com/xmake-io/xmake/blob/4d20b980c0d3826641a7b0481145a5e495417f6e/xmake/rules/verilator/xmake.lua#L39

https://github.com/xmake-io/xmake/blob/4d20b980c0d3826641a7b0481145a5e495417f6e/xmake/rules/verilator/verilator.lua#L210-L265

https://github.com/xmake-io/xmake/blob/4d20b980c0d3826641a7b0481145a5e495417f6e/xmake/modules/private/action/build/object.lua#L169

https://github.com/xmake-io/xmake/issues/6202

Reply:
I have converted the code to use jobgraph